src/zonetree/answer.rs

```diff
@@ -0,0 +1,217 @@
+///
+/// [`Answer`] is the type returned by [`ReadableZone::query()`].
+///
+/// Callers of [`ReadableZone::query()`] will likely only ever need to use the
+/// [`Self::to_message()`] function. Alternatively, for complete control use
+/// the getter functions on [`Answer`] instead and construct a response
+/// message yourself using [`MessageBuilder`].
+///
+/// Implementers of alternate backing stores for [`Zone`]s will need to use
+/// one of the various `Answer` constructor functions when
+/// [`ReadableZone::query()`] is invoked for your zone content in order to
+/// tailor the DNS message produced by [`Self::to_message()`] based on the
+/// outcome of the query.
+///
+/// [`Zone`]: crate::zonetree::Zone
+/// [`ReadableZone::query()`]: crate::zonetree::traits::ReadableZone::query()
+#[derive(Clone)]
+pub struct Answer {
+    /// The response code of the answer.
+    rcode: Rcode,
+
+    /// The content of the answer.
+    content: AnswerContent,
+
+    /// The optional authority section to be included in the answer.
+    authority: Option<AnswerAuthority>,
+}
+
+impl Answer {
+    /// Creates an "empty" answer.
+    ///
+    /// The answer, authority and additinal sections will be empty.
+    pub fn new(rcode: Rcode) -> Self {
+        Answer {
+            rcode,
+            content: AnswerContent::NoData,
+            authority: Default::default(),
+        }
+    }
+
+    /// Creates a new message with a populated authority section.
+    ///
+    /// The answer and additional sections will be empty.
+    pub fn with_authority(rcode: Rcode, authority: AnswerAuthority) -> Self {
+        Answer {
+            rcode,
+            content: AnswerContent::NoData,
+            authority: Some(authority),
+        }
+    }
+
+    /// Creates a new [Rcode::Refused] answer.
+    ///
+    /// This is equivalent to calling [`Answer::new(Rcode::Refused)`].
+    pub fn refused() -> Self {
+        Answer::new(Rcode::Refused)
+    }
+
+    /// Adds a CNAME to the answer section.
+    pub fn add_cname(&mut self, cname: SharedRr) {
+        self.content = AnswerContent::Cname(cname);
+    }
+
+    /// Adds an RRset to the answer section.
+    pub fn add_answer(&mut self, answer: SharedRrset) {
+        self.content = AnswerContent::Data(answer);
+    }
+
+    /// Sets the content of the authority section.
+    pub fn set_authority(&mut self, authority: AnswerAuthority) {
+        self.authority = Some(authority)
+    }
+
+    /// Prepares a response to the given request based on the properties of
+    /// this answer.
+    ///
+    /// The response [Rcode], question, answer and authority sections of the
+    /// produced [`AdditionalBuilder`] will be populated based on the
+    /// properties of this [`Answer`] as determined by the constructor and
+    /// add/set functions called prior to calling this function.
+    ///
+    /// See also: [`MessageBuilder::start_answer()`]
+    pub fn to_message<RequestOctets: Octets, Target: Composer>(
```

Comment:
It is not clear to me where the AA flag gets set

Reply:
It doesn't get set. See https://github.com/NLnetLabs/domain/pull/286#discussion_r1553417085 above.